src/librustc/traits/object_safety.rs

```diff
@@ -323,9 +316,108 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
             return Some(MethodViolationCode::WhereClauseReferencesSelf(span));
         }
 
+        let receiver_ty = self.liberate_late_bound_regions(
+            method.def_id,
+            &sig.map_bound(|sig| sig.inputs()[0]),
+        );
+
+        // until we get by-value DST, `self: Self` can't be coerced
+        // but we allow this as a special case.
+        // maybe instead we should also check for
+        // for (U) { if (Self: Unsize<U>) { Receiver: Unsize<Receiver<Self=Self>>}}
+        if receiver_ty != self.mk_self_type() {
+            if !self.receiver_is_coercible(method, receiver_ty) {
+                return Some(MethodViolationCode::UncoercibleReceiver);
+            }
+        }
+
         None
     }
 
+    // checks the type of the self argument, and makes sure it implements
+    // the CoerceUnsized requirement:
+    // forall (U) {
+    //     if (Self: Unsize<U>) {
+    //         Receiver: CoerceUnsized<Receiver<Self=U>>
+    //     }
+    // }
+    #[allow(dead_code)]
+    fn receiver_is_coercible(
+        self,
+        method: &ty::AssociatedItem,
+        receiver_ty: Ty<'tcx>,
+    ) -> bool
+    {
+        debug!("receiver_is_coercible: method = {:?}, receiver_ty = {:?}", method, receiver_ty);
+
+        let traits = (self.lang_items().unsize_trait(),
+                      self.lang_items().coerce_unsized_trait());
+        let (unsize_did, coerce_unsized_did) = if let (Some(u), Some(cu)) = traits {
+            (u, cu)
+        } else {
+            debug!("receiver_is_coercible: Missing Unsize or CoerceUnsized traits");
+            return false;
+        };
+
+        // use a bogus type parameter to mimick a forall(U) query
```

Comment:
yea plz do something else here

Reply:
haha sure. I'm not sure what the cleanest way to do this is, but I'll ping Niko on Discord next week and see what he says.

Reply:
Yeah, it's a good question how to do this cleanly. We encountered this with chalk integration and kind of "kicked the bucket", but maybe it's worth trying to fix properly (we might consider landing this hack -- perhaps with some edits to `ppaux`, and probably changed to `RustaceansRock`, of course 😉 -- and then do follow-up work to fix properly (@mikeyhew the refactorings to debruijn indices were partly motivated by this as well, actually)